### ISSUE TITLE
Declaring queues on worker start

### ISSUE THREAD

**Issue by xelhark** [celery/celery#5375]
I have a very time-sensitive application that I'm building.

I noticed that the first time a worker sends a new task, it has a non-trivial delay due to the declaring of the queues.

I would like to know if it's possible to remove this delay, or to move it at some init function in my worker.

## Related Issues and Possible Duplicates

None found

#### Related Issues

None found

#### Possible Duplicates

- None

## Environment & Settings
**Celery version**: 4.2.1 (windowlicker)
<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
software -> celery:4.2.1 (windowlicker) kombu:4.4.0 py:3.6.6
            billiard:3.5.0.5 redis:3.2.0
platform -> system:Darwin arch:64bit imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:disabled

broker_url: 'redis://localhost:6379//'
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies
* **Minimal Python Version**: N/A or Unknown
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: N/A or Unknown
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages
<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
amqp==2.4.2
billiard==3.5.0.5
celery==4.2.1
kombu==4.4.0
pytz==2018.9
redis==3.2.0
vine==1.2.0
```

</p>
</details>

### Other Dependencies
<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case
<details>
<p>

```python
import time

from celery import Celery

app = Celery('test', broker='redis://localhost:6379')


@app.task()
def dummy_task_secondary():
    print("Dummy task completed")


@app.task()
def dummy_task_primary():
    t = time.time()
    dummy_task_secondary.delay()
    print(f"First time it took {time.time() - t}")

    t = time.time()
    dummy_task_secondary.delay()
    print(f"Second time it took {time.time() - t}")

```

</p>
</details>

# Expected Behavior
There is no difference between the time it takes to send the first task and the second.

# Actual Behavior
The first time is dramatically slower:

```
[2019-03-11 15:27:42,610: WARNING/ForkPoolWorker-8] First time it took 0.029098987579345703
[2019-03-11 15:27:42,612: WARNING/ForkPoolWorker-8] Second time it took 0.0007290840148925781
```

**Comment by othalla** [comment]
@xelhard why don't you manually create some uniq queue to solve this problem and route all tasks to it? 
This may be a quick fix

**Comment by xelhark** [comment]
Could you elaborate? How would I be able to do that? Maybe some example configuration code would help 😄

**Comment by othalla** [comment]
Humm to be honest i've tested my solution and it the result is the same. I suspect that kombu queue is created when the first task is proceed.
I'll check this out deeper!

**Comment by xelhark** [comment]
>  I suspect that kombu queue is created when the first task is proceed.

I also think that's the case. From what I've been able to see in the code it's the `kombu.common.maybe_declare` function that takes most of that time.

I'd love to be able to do it when loading the worker, it would really help in speeding up performances. With remote brokers that declaration time ends up being even ~400 ms which is a lot for some applications.

**Comment by othalla** [comment]
You still have to wait for the worker process to start up which is quite slow

**Comment by xelhark** [comment]
That's not a problem, I would start it way earlier but then I'd know that once it's ready there are no more delays 😄

**Comment by othalla** [comment]
Okay, i'll have a look and see what can be done

**Comment by xelhark** [comment]
Any news about this?

**Comment by othalla** [comment]
Hey, i'm still checking inside the code. I was quite busy this time.

**Comment by xelhark** [comment]
No worries. Let me know if you have insights or if I can help, I'd be glad to contribute.

**Comment by othalla** [comment]
@xelhark i started to search if i can do something but you can search too if you want. This is not easy to dig inside celery code :)

**Comment by othalla** [comment]
From what i've checked, the first task sent on some child process processing the queue trigger the connection which is required for producing kombu messages.
I keep digging!

**Comment by xelhark** [comment]
Sooo, I've been digging a bit on my own on this, and this is what I've found:

The Kombu function to declare the queues is responsible for that delay. 
The workers themselves do not declare the queues (probably because it's not a common use case to send a task from another task, and the delay is not that big anyway), so this happens.

I found a way to pre-declare the queues, and this seems to be working locally:
Here's the code:

```
# Preload AMPQ queues in celery
with app.producer_or_acquire() as producer:
    for queue in app.ampq.queues:
        producer.maybe_declare(queue)
```

When adding this code to the app config file, the delay does not happen anymore (locally).

Unfortunately, this doesn't work with a regular worker, so if I launch a worker with `celery -A myapp worker`, this initialization code won't be run on that worker init, so it will still declare the queues lazily.

I also tried to use a custom worker class and use that code as an initialization code in the `init` of my worker class, but again, it seems like it's not working 😞  

Do you have any clues about why?

**Comment by othalla** [comment]
what do you mean by working locally?

**Comment by xelhark** [comment]
I mean in Eager mode (so this won't work with a worker declared on another task, but it will make the delay disappear if you attempt to run it locally).

**Comment by thedrow** [comment]
I suggest that you'll trying using a [bootstep](http://docs.celeryproject.org/en/latest/userguide/extending.html) for now.

**Comment by othalla** [comment]
Tried with a bootstep, looks like it works.

**Comment by xelhark** [comment]
I'm sorry, but could you add a code example? I tested it and it doesn't seem to be working, maybe I'm doing something wrong?

```python
import time

from celery import Celery, bootsteps

app = Celery('test', broker='redis://localhost:6379')


class PreloadBoostep(bootsteps.StartStopStep):
    def start(self, parent):
        print(f"Preloading worker...")
        app = parent.app
        with app.producer_or_acquire() as producer:
            for queue in app.amqp.queues.values():
                print(f"Preloading queue: {queue}")
                producer.maybe_declare(queue)


app.steps['worker'].add(PreloadBoostep)


@app.task()
def dummy_task_secondary():
    print("Dummy task completed")


@app.task()
def dummy_task_primary():
    t = time.time()
    dummy_task_secondary.delay()
    print(f"First time it took {time.time() - t}")

    t = time.time()
    dummy_task_secondary.delay()
    print(f"Second time it took {time.time() - t}")

```

From the logs:

```
[2019-04-08 11:31:46,038: WARNING/MainProcess] Preloading worker...
[2019-04-08 11:31:46,047: WARNING/MainProcess] Preloading queue: <unbound Queue celery -> <unbound Exchange celery(direct)> -> celery>
[2019-04-08 11:31:55,907: WARNING/ForkPoolWorker-8] First time it took 0.03023982048034668
[2019-04-08 11:31:55,908: WARNING/ForkPoolWorker-8] Second time it took 0.0008070468902587891
[2019-04-08 11:31:55,909: WARNING/ForkPoolWorker-9] Dummy task completed
[2019-04-08 11:31:55,911: WARNING/ForkPoolWorker-10] Dummy task completed
```

It seems to me that the preloading actually did happen (so the bootstep worked) but in the end it still took a lot more time to send the first task than it did for the second.

**Comment by xelhark** [comment]
@othalla  @thedrow  are there any news about this?

**Comment by othalla** [comment]
You're right, after digging a bit, seems it does not solve the problem. I'll check this out deeper and give you some feedback.

**Comment by xelhark** [comment]
Any news on this?

**Comment by thedrow** [comment]
Since you declared the queues in the bootstep, can you verify that they are being created in RabbitMQ?
If so, our problem is elsewhere.

**Comment by xelhark** [comment]
> Since you declared the queues in the bootstep, can you verify that they are being created in RabbitMQ? If so, our problem is elsewhere.
What do you mean?
I included some code in the minimally reproducible test case, I'm currently using Redis as a broker (not RabbitMQ) so I can't say if the queues are created.

The code is actually very easy to execute, it's just a few lines.

**Comment by thedrow** [comment]
So maybe the problem is how we create the queues in Redis or how we query that the queue exists for the first time before inserting a message into it?

**Comment by xelhark** [comment]
Honestly, I don't know, I just noticed that there is this big delay when adding a message for the first time in a new queue.

Here's a very quick script to check it out:

```
import time

from celery import Celery

app = Celery('test', broker='redis://localhost:6379')


@app.task()
def dummy_task_secondary():
    print("Dummy task completed")


@app.task()
def dummy_task_primary():
    t = time.time()
    dummy_task_secondary.delay()
    print(f"First time it took {time.time() - t}")

    t = time.time()
    dummy_task_secondary.delay()
    print(f"Second time it took {time.time() - t}")
```

I have a very time-sensitive application and when this happens in my server, with a redis remote instance that first `.delay()` can take up to 200 ms which is a lot of time in my app.

I have been forced to use a very ugly workaround, it would be great to have celery working properly here. I think I provided enough information, as I said earlier, the core of the time spent seems to be in the `kombu.common.maybe_declare` function that "declares" the queue itself in redis.

Except I don't know what is actually taking so much time, it doesn't have to write any data in there, does it?
